File: tools/cross-validation/parser-ts.js
/**
 * Parser for tree-sitter S-expression output.
 * @module parser-ts
 */

'use strict';

const { Node } = require('./models');

/**
 * Regex to match position annotations like [0, 0] - [7, 0]
 * @type {RegExp}
 */
const POSITION_RE = /\[\d+, \d+\] - \[\d+, \d+\]/g;

/**
 * Tokenize S-expression string into tokens: '(', ')', and word tokens.
 * @param {string} text - Cleaned S-expression text (positions stripped).
 * @returns {string[]} Array of tokens.
 */
function tokenize(text) {
  const tokens = [];
  let i = 0;
  while (i < text.length) {
    const ch = text[i];
    if (ch === '(' || ch === ')') {
      tokens.push(ch);
      i++;
    } else if (/\s/.test(ch)) {
      i++;
    } else {
      // Word token
      let start = i;
      while (i < text.length && text[i] !== '(' && text[i] !== ')' && !/\s/.test(text[i])) {
        i++;
      }
      tokens.push(text.slice(start, i));
    }
  }
  return tokens;
}

/**
 * Parse tokens into a Node tree recursively.
 * @param {string[]} tokens - Token array.
 * @param {{pos: number}} state - Mutable parsing state.
 * @returns {Node} Parsed node.
 */
function parseNode(tokens, state) {
  // Expect '('
  if (tokens[state.pos] !== '(') {
    throw new Error(`Expected '(' at position ${state.pos}, got '${tokens[state.pos]}'`);
  }
  state.pos++; // consume '('

  // Read node name
  const name = tokens[state.pos];
  state.pos++;

  // Read children until ')'
  const children = [];
  while (state.pos < tokens.length && tokens[state.pos] !== ')') {
    if (tokens[state.pos] === '(') {
      children.push(parseNode(tokens, state));
    } else {
      // Skip unexpected non-paren tokens (shouldn't normally happen)
      state.pos++;
    }
  }

  // Expect ')'
  if (tokens[state.pos] === ')') {
    state.pos++; // consume ')'
  }

  return new Node(name, children);
}

/**
 * Parse tree-sitter S-expression output into a Node tree.
 * @param {string} text - The S-expression text from tree-sitter.
 * @returns {Node} The root node.
 */
function parseTreeSitter(text) {
  // Strip position annotations
  const cleaned = text.replace(POSITION_RE, '');

  const tokens = tokenize(cleaned);
  if (tokens.length === 0) {
    return new Node('source_file');
  }

  const state = { pos: 0 };
  return parseNode(tokens, state);
}

module.exports = { parseTreeSitter };

File: tools/cross-validation/parser_ts.py
"""Parser for tree-sitter S-expression output format.

Parses output like:
    (source_file [0, 0] - [7, 0]
      (class_declaration [4, 0] - [6, 1]
        (type_identifier [4, 6] - [4, 14])))

Into a tree of Node objects, stripping position info and keeping only named nodes.
"""

from __future__ import annotations

import re
from models import Node


def parse_tree_sitter(text: str) -> Node:
    """Parse a tree-sitter S-expression string into a Node tree.

    Args:
        text: The S-expression output from `tree-sitter parse`.

    Returns:
        The root Node of the parsed tree.

    Raises:
        ValueError: If the input is empty or malformed.
    """
    text = text.strip()
    if not text:
        raise ValueError("Empty input")

    tokens = _tokenize(text)
    node, _ = _parse_node(tokens, 0)
    return node


def _tokenize(text: str) -> list[str]:
    """Tokenize an S-expression into a list of meaningful tokens.

    Splits the text into '(', ')', node names, and skips position annotations
    like '[0, 0] - [7, 0]'.
    """
    # Remove position annotations: [row, col] - [row, col]
    cleaned = re.sub(r"\[\d+,\s*\d+\]\s*-\s*\[\d+,\s*\d+\]", "", text)

    tokens: list[str] = []
    i = 0
    while i < len(cleaned):
        ch = cleaned[i]
        if ch in " \t\n\r":
            i += 1
            continue
        if ch == "(":
            tokens.append("(")
            i += 1
            continue
        if ch == ")":
            tokens.append(")")
            i += 1
            continue
        # Read a node name (word characters, underscores, dots)
        j = i
        while j < len(cleaned) and cleaned[j] not in "() \t\n\r":
            j += 1
        word = cleaned[i:j]
        if word:
            tokens.append(word)
        i = j
    return tokens


def _parse_node(tokens: list[str], pos: int) -> tuple[Node, int]:
    """Parse a single node starting at the given token position.

    A node is: '(' name children... ')'

    Returns:
        A tuple of (parsed Node, next token position).
    """
    if pos >= len(tokens) or tokens[pos] != "(":
        raise ValueError(f"Expected '(' at position {pos}, got {tokens[pos] if pos < len(tokens) else 'EOF'}")

    pos += 1  # skip '('

    if pos >= len(tokens):
        raise ValueError("Unexpected end of input after '('")

    name = tokens[pos]
    pos += 1

    children: list[Node] = []
    while pos < len(tokens) and tokens[pos] != ")":
        if tokens[pos] == "(":
            child, pos = _parse_node(tokens, pos)
            children.append(child)
        else:
            # Skip any unexpected tokens (shouldn't happen after cleaning)
            pos += 1

    if pos >= len(tokens):
        raise ValueError("Unexpected end of input, expected ')'")

    pos += 1  # skip ')'
    return Node(name=name, children=children), pos
